enhancements/proposed/20200501-json-output/start-json-output.md

```diff
@@ -0,0 +1,177 @@
+This allows end users to see clear, and ideally actionable, error messages when minikube breaks.
+
+
+
+## Goals
+
+### Stderr
+*   Error code from [err_map.go](https://github.com/kubernetes/minikube/blob/master/pkg/minikube/problem/err_map.go) is parsable from stderr
+*   In case of a panic, default to sending all logs to stderr so that the user can see all logs
+
+### Stdout
+*   Progress is communicated with the total numbers of steps and the current step the user is on
+*   An encoded step name, like `pull_images`, would be useful here
+
+
+## Non-Goals
+
+*   Improving error handling in minikube; this is just about how we output errors to users
+
+## Design Details
+Users can specify JSON output on minikube start via a flag:
+
+```
+minikube start --output json
+```
+
+### Stderr
+Logs can be sent to stderr as usual.
+In addition, we will output error code from [err_map.go](https://github.com/kubernetes/minikube/blob/master/pkg/minikube/problem/err_map.go) as a parsable JSON message to stderr.
+
+This will be done by adding the following function to the `out` package:
+
+```go
+func DisplayErrorJSON(out io.Writer, p *problem.Problem)
+```
+
+which will print the JSON encoding of `p` to `out`.
+
+If JSON output is specified, and minikube finds an applicable error message in `err_map.go`, then we can call `DisplayErrorJSON` when handling the error in the [WithError](https://github.com/kubernetes/minikube/blob/master/pkg/minikube/exit/exit.go#L57) function:
+
+```go
+// WithError outputs an error and exits.
+func WithError(msg string, err error) {
+	...
+	p := problem.FromError(err, runtime.GOOS)
+	if p != nil {
+        WithProblem(msg, err, p)
+        // Add this function here (this is just pseudocode)
+        if json {
+            out.DisplayErrorJSON(os.Stderr, p)
+        }
+	}
+    ...
+}
+```
+
+#### Testing Plan
+This feature will be covered by unit tests exclusively.
+If a problem is found, and json is specified, then we just want to make sure that the JSON output is parsable.
+
+### Stdout
+As mentioned above, the requirements for stdout are:
+1. Steps are numbered, and the total number of steps is known
+1. Steps have a helpful name
+
+First, we'll need some way of distinguishing between logs.
+Each log has three components:
+1. The message itself 
+1. The emoji associated with the log (StyleEnum)
+1. Whether the log is of type Log, Warning, or Error
+
+```go
+type Log struct {
+	LogType // Will be either Log, Warning or Error
+	style   StyleEnum
+	message string
+}
+```
+
+
+We also need to know the total number of steps before minikube starts.
+For that to be possible, we need to know all of the logs we will print before starting.
+
+I propose creating a registry of logs, which is prefilled with all necessary logs before we start minikube.
+The registry would look like this:
+
+```go
+// Registry holds all user-facing logs
```

Comment:
In general, I would prefer that implementation code & structs be kept out of a design proposal. Designs should focus on the novel interesting parts of the design.

For an API-focused change like this, it's fine to mention what the new API would look like, however. Comments  in particular should not be included.